clang/lib/Sema/SemaSYCL.cpp

```diff
@@ -1404,13 +1391,13 @@ static bool isKnownEmitted(Sema &S, FunctionDecl *FD) {
   if (!FD)
     return true; // Seen in LIT testing
 
-  if (FD->hasAttr<SYCLDeviceAttr>() || FD->hasAttr<SYCLKernelAttr>())
```

Comment:
Because otherwise we end up double diagnosing.  The template part has to be first, otherwise SCYLDevice marked templates will diagnose both when going through stage 1 and then when being instantiated.

Reply:
Am I right that the second diagnosing happens in SemaSYCL when we marked all device functions with `SYCLDevice` attribute?

Reply:
No, only the emission of the diagnostic happens when functions are marked with the device attribute.  The repro I had was something like (from memory, pardon the typos):

    <template>
    __attribute__((sycldevice)) void some_func() {
        variadic_call();
   }

In this case, during phase 1 (when the template is being parsed, but is uninstantiated), this would be diagnosed  immediately thanks to the marked line being above the template check.

THEN, it would be called (instantiated), and would diagnose AGAIN there.  THEN, in cases where it was the actual kernel, we would diagnose a THIRD time (since it was the "OpenCLKernel"!).  

Not that my changes to return K_Nop in the function below might make the template check here not necessary, but I wasn't positive so I left it.

Reply:
Ok, thanks for the explanation.